table/substrait/substrait.go

```diff
@@ -0,0 +1,379 @@
+func (t *toSubstraitExpr) VisitOr(left, right expr.Builder) expr.Builder {
+	return t.bldr.ScalarFunc(orID).Args(left.(expr.FuncArgBuilder),
+		right.(expr.FuncArgBuilder))
+}
+
+func (t *toSubstraitExpr) VisitUnbound(iceberg.UnboundPredicate) expr.Builder {
+	panic("can only convert bound expressions to substrait")
+}
+
+func (t *toSubstraitExpr) VisitBound(pred iceberg.BoundPredicate) expr.Builder {
+	return iceberg.VisitBoundPredicate(pred, t)
+}
+
+type substraitPrimitiveLiteralTypes interface {
+	bool | ~int32 | ~int64 | float32 | float64 | string
+}
+
+func toPrimitiveSubstraitLiteral[T substraitPrimitiveLiteralTypes](v T) expr.Literal {
+	return expr.NewPrimitiveLiteral(v, false)
+}
+
+func toByteSliceSubstraitLiteral[T []byte | types.UUID](v T) expr.Literal {
+	return expr.NewByteSliceLiteral(v, false)
+}
+
+func toDecimalLiteral(v iceberg.DecimalLiteral) expr.Literal {
+	byts, _ := v.MarshalBinary()
+	result, _ := expr.NewLiteral(&types.Decimal{
+		Scale:     int32(v.Scale),
+		Value:     byts,
+		Precision: int32(v.Type().(*iceberg.DecimalType).Precision()),
+	}, false)
+	return result
+}
+
+func toFixedLiteral(v iceberg.FixedLiteral) expr.Literal {
+	return expr.NewFixedBinaryLiteral(types.FixedBinary(v), false)
+}
+
+func toSubstraitLiteral(typ iceberg.Type, lit iceberg.Literal) expr.Literal {
+	switch lit := lit.(type) {
+	case iceberg.BoolLiteral:
+		return toPrimitiveSubstraitLiteral(bool(lit))
+	case iceberg.Int32Literal:
+		return toPrimitiveSubstraitLiteral(int32(lit))
+	case iceberg.Int64Literal:
+		return toPrimitiveSubstraitLiteral(int64(lit))
+	case iceberg.Float32Literal:
+		return toPrimitiveSubstraitLiteral(float32(lit))
+	case iceberg.Float64Literal:
+		return toPrimitiveSubstraitLiteral(float64(lit))
+	case iceberg.StringLiteral:
+		return toPrimitiveSubstraitLiteral(string(lit))
+	case iceberg.TimestampLiteral:
+		if typ.Equals(iceberg.PrimitiveTypes.TimestampTz) {
+			return toPrimitiveSubstraitLiteral(types.TimestampTz(lit))
+		}
+		return toPrimitiveSubstraitLiteral(types.Timestamp(lit))
+	case iceberg.DateLiteral:
+		return toPrimitiveSubstraitLiteral(types.Date(lit))
+	case iceberg.TimeLiteral:
+		return toPrimitiveSubstraitLiteral(types.Time(lit))
+	case iceberg.BinaryLiteral:
+		return toByteSliceSubstraitLiteral([]byte(lit))
+	case iceberg.FixedLiteral:
+		return toFixedLiteral(lit)
+	case iceberg.UUIDLiteral:
+		return toByteSliceSubstraitLiteral(types.UUID(lit[:]))
+	case iceberg.DecimalLiteral:
+		return toDecimalLiteral(lit)
+	}
+	panic(fmt.Errorf("invalid literal type: %s", lit.Type()))
+}
+
+func toSubstraitLiteralSet(typ iceberg.Type, lits []iceberg.Literal) expr.ListLiteralValue {
+	if len(lits) == 0 {
+		return nil
+	}
+
+	out := make([]expr.Literal, len(lits))
+	for i, l := range lits {
+		out[i] = toSubstraitLiteral(typ, l)
+	}
+	return out
+}
+
+func (t *toSubstraitExpr) getRef(ref iceberg.BoundReference) expr.Reference {
+	updatedRef, err := iceberg.Reference(ref.Field().Name).Bind(t.schema, true)
```

Comment:
Do we want to pass down the case-sensitivity from the ScanOption?

Reply:
good point, i've updated everything to pass the caseSensitive down from the ScanOption